src/neuroconv/basedatainterface.py

```diff
@@ -46,6 +47,91 @@ def get_metadata(self):
+        )
+
+    @abstractmethod
+    def get_timestamps(self) -> np.ndarray:
+        """
+        Retrieve the timestamps for the data in this interface.
+
+        Returns
+        -------
+        timestamps: numpy.ndarray
+            The timestamps for the data stream.
+        """
+        raise NotImplementedError(
+            "Unable to retrieve timestamps for this interface! Define the `get_timestamps` method for this interface."
+        )
+
+    @abstractmethod
+    def align_timestamps(self, aligned_timestamps: np.ndarray):
+        """
+        Replace all timestamps for this interface with those aligned to the common session start time.
+
+        Must be in units seconds relative to the common 'session_start_time'.
+
+        Parameters
+        ----------
+        aligned_timestamps : numpy.ndarray
+            The synchronized timestamps for data in this interface.
+        """
+        raise NotImplementedError(
+            "The protocol for synchronizing the timestamps of this interface has not been specified!"
+        )
+
+    def align_starting_time(self, starting_time: float):
+        """
+        Align the starting time for this interface relative to the common session start time.
+
+        Must be in units seconds relative to the common 'session_start_time'.
+
+        Parameters
+        ----------
+        starting_time : float
+            The starting time for all temporal data in this interface.
+        """
+        self.align_timestamps(aligned_timestamps=self.get_timestamps() + starting_time)
+
+    def align_by_interpolation(self, unaligned_timestamps: np.ndarray, aligned_timestamps: np.ndarray):
+        """
+        Interpolate the timestamps of this interface using a mapping from some unaligned time basis to its aligned one.
+
+        Use this method if the unaligned timestamps of the data in this interface are not directly tracked by a primary
+        system, but are known to occur between timestamps that are tracked, then align the timestamps of this interface
+        by interpolating between the two.
+
+        An example could be a metronomic TTL pulse (e.g., every second) from a secondary data stream to the primary
+        timing system; if the time references of this interface are recorded within the relative time of the secondary
+        data stream, then their exact time in the primary system is inferred given the pulse times.
+
+        Must be in units seconds relative to the common 'session_start_time'.
+
+        Parameters
+        ----------
+        unaligned_timestamps : numpy.ndarray
+            The timestamps of the unaligned secondary time basis.
+        aligned_timestamps : numpy.ndarray
+            The timestamps aligned to the primary time basis.
+        """
+        self.align_timestamps(
+            aligned_timestamps=np.interp(x=self.get_timestamps(), xp=unaligned_timestamps, fp=aligned_timestamps)
+        )
```

Comment:
I think the term "timestamps" is a bit overloaded here, because it refers to the sample times of the signal we are trying to align and to the times of the ttl pulses. What do you think about renaming:

unaligned_timestamps -> ttl_sent_times
aligned_timestamps -> ttl_received_times

or something like that

Reply:
`aligned_timestamps` is consistent with the keyword argument for `align_timestamps`, which is consistent with its usage in adjusting the `timestamps` field of the underlying object (when the object is a `TimeSeries`, at least)

Keep in mind the `timestamps` also may not be inferred only through TTL signals - it is also possible to decode the timings via more complicated digital words (compound TTL, each sent at different times and parsed together) or perhaps even some other mechanism (like recording the system time in UNIX seconds and saving it in a CSV)

So I'd say the argument names ought to be agnostic to the 'how' of the timings

If `timestamps` is too long, we can mimic SpikeInterface/ROIExtractors by just referring to it always as `times`

```
get_timestamps -> get_times
align_timestamps -> align_times
unaligned_timestamps -> unaligned_times
aligned_times -> aligned_times
```

in particular, `get_times` would be eventually refactored into that enhanced treatment of `starting_time` + `rate` vs. `timestamps` that will be eventually added to the write tools